design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,121 @@
+
+* Aggregates should not be allowed (please see *[Open questions](#open-questions)* section)
+* `SELECT` subqueries should not be allowed
+
+#### Examples
+
+Given these restrictions, the following examples would be valid lambda functions:
+
+* `x => x + 1`
+* `s => UCASE(s)`
+* `(x, y) => x + y`
+* `(x, y) => CASE WHEN x IS NULL THEN y ELSE x + y END`
+* `(x, y) => x + COALESCE(y, 0)`
+
+### Invocation functions
+
+Lambda functions should require a specific **invocation function**, which tells ksqlDB how to apply the given lambda function to an input value, and what kind of result to return. The following invocation functions are proposed for the initial support for lambda functions:
+
+* `transform_array(arr, x => y)` - Applies the given lambda function to each element of the input `Array`, returning a new `Array` containing the transformed output.
+
+* `transform_map(map, (k, v) => x, y)` - Applies the given lambda function to each key-value pair of the input `Map`, returning a new `Map` containing the transformed key-value pairs.
+
+* `filter_array(arr, x => bool)` - Filters the input `Array` using the given lambda function. A new `Array` is returned, containing only values for which the lambda function evaluated to `true`.
+
+* `filter_map(map, (k, v) => bool)` - Filters the input `Map` using the given lambda function. A new `Map` is returned, containing only the key-value pairs for which the lambda function evaluated to `true`.
+
+* `reduce_array(arr, s, (x, s) => s)` - Reduces the input `Array` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_array` will return the final value of `s`.
+
+* `reduce_map(map, s, (k, v, s) => s)` - Reduces the input `Map` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_map` will return the final value of `s`.
+
+## Documentation updates
+
+We must document and provide examples for the two core aspects of lambda functions:
+
+1. Lambda function syntax
+2. Invocation functions
+
+## Compatibility implications
+
+Lambda functions would be additive to ksqlDB's grammar and should therefore not introduce any incompatibilities into the grammar or public APIs. However, it is not currently known if a lambda function implementation would require backward-incompatible changes to ksqlDB's internals.
+
+## Security implications
+
+Since lambda functions would only allow SQL expressions that can already be used within queries, no new security implications should be introduced.
+
+## Open questions
+
+**1. Would it be possible to allow aggregates in lambdas within the context of a particular invocation function? For example,**
+
+- `agg_array(arr, x => sum(x))`
```

Comment:
This seems a little complex for something you could implement as `reduce_array(arr, (x, y) => x + y)`.  How do you imagine this being used with `reduce_*` functions?  This seems like it would be a separate replacement  for functional reduce, with labmdas implemented a little differently.

Reply:
`sum` is actually not the best example here. What about aggregates with non-scalar intermediate state, such as `avg`?

Reply:
The more I think about this, the more it feels like using a builtin aggregate expression within a lambda function doesn't really semantically make sense. I think we could theoretically do it but the example here isn't technically functional so it feels wrong. e.g there is nothing passed into the lambda that would tell each invocation where the last one left off.

I think that if we wanted to support aggregations over collections, we'd want to do it with some kind of `apply` function:
```sql
SELECT apply_array(arr, sum) ...
SELECT apply_array(arr, avg) ...
```

Subqueries would probably be more SQL-like:
```sql
SELECT (SELECT sum(explode(arr))) ...
```

I don't think either of these should be within the scope of this KLIP though.